uwblueprint/focus-on-nature
# Setup Stripe webhooks

[1/2] https://www.notion.so/uwblueprintexecs/Task-Board-F22-9e03bf9f7c4343f09a8a5c16dc48a12a?p=7f51d5e490d0494ea2512dd2b463186d&pm=c

- Need to fix the camper charges if there was a coupon used


## Implementation description
* Right now, we send confirmation email in the "create campers" request, and then confirm payment as a separate request. The confirmation email (as well as sending the waiver eventually) should be sent after payment is completed
  * We do have to figure out how to handle incomplete checkouts -- should they continue holding capacity in the camp forever?
* WHEN THIS GETS MERGED TO STAGING: Enable the backend endpoint in the [test Stripe dashboard](https://dashboard.stripe.com/test/webhooks)
* WHEN THIS GETS MERGED TO MAIN: Add the backend endpoint in the production Stripe dashboard to listen to `checkout.session.completed` events (I think FON has credentials, I don't have access)



## Steps to test
1. Go through payments flow, check that you've (1) received the email, (2) it's been marked as `hasPaid` in Mongo. For local testing, download the Stripe CLI, `stripe login` and forward using `stripe listen --forward-to localhost:5000/campers/confirm-payment`

`stripe trigger checkout.session.completed` is another useful test command


## What should reviewers focus on?
* Anyone have load concerns?
* Various docs to know:
  * https://stripe.com/docs/webhooks
  * https://stripe.com/docs/webhooks/quickstart
  * https://stripe.com/docs/cli/trigger#trigger-event
  * https://stripe.com/docs/api/checkout/sessions/object
  * https://stripe.com/docs/api/events/object


## Checklist
